Re-use ImportThunks in Crossgen2

nattress opened the pull request:
Reduce the number of thunks generated to be in line with Crossgen1. Each unique method that is called currently gets its own thunk used for that specific import section entry. Change the thunk calling pattern to match Crossgen1, allowing a small handful of thunks to be re-used for all method / virtual method calls.

* Create ImportThunks using the `NodeFactory` instead of allocating directly to allow sharing between different imports.

* On x64, switch method call helper thunk to use following assembly:
```
xor eax, eax
push index
push [module]
jmp [helperCell]
```

* Previously we emitted this assembly for the thunks:
```
lea eax, [instanceCell]
push index
push [module]
jmp [helperCell]
```

* The runtime doesn't need the import cell for the call passing in - it finds it by disassembling the address from `call [helper]`.

* Each thunk encodes the containing import section index; duplication of thunks is thus allowed per import section.

* Share virtual method call fixup cells between all invocations of a given virtual method. Currently, each unique method calling a given virtual method gets a fixup. That can have a runtime performance gain but comes at the expense of binary size. Match Crossgen1 behavior and use a single fixup cell per virtual method.

nattress committed:
Re-use ImportThunks in Crossgen2

Reduce the number of thunks generated to be in line with Crossgen1. Each unique method that is called currently gets its own thunk used for that specific import section entry. Change the thunk calling pattern to match Crossgen1, allowing a small handful of thunks to be re-used for all method / virtual method calls.

* Create ImportThunks using the `NodeFactory` instead of allocating directly to allow sharing between different imports.

* On x64, switch method call helper thunk to use following assembly:
```
xor eax, eax
push index
push [module]
jmp [helperCell]
```

* Previously we emitted this assembly for the thunks:
```
lea eax, [instanceCell]
push index
push [module]
jmp [helperCell]
```

* The runtime doesn't need the import cell for the call passing in - it finds it by disassembling the address from `call [helper]`.

* Each thunk encodes the containing import section index; duplication of thunks is thus allowed per import section.

* Share virtual method call fixup cells between all invocations of a given virtual method. Currently, each unique method calling a given virtual method gets a fixup. That can have a runtime performance gain but comes at the expense of binary size. Match Crossgen1 behavior and use a single fixup cell per virtual method.

MichalStrehovsky commented on src/coreclr/tools/Common/Compiler/DependencyAnalysis/Target_X64/X64Emitter.cs:
```diff
@@ -174,6 +174,15 @@ public void EmitRETIfEqual()
             Builder.EmitByte(0xC3);
         }
 
+        public void EmitXOR(Register regDst, Register regSrc)
```
If we want to call this EmitXOR, it should be something along the lines of:

```csharp
public void EmitXOR(Register regDst, ref AddrMode memory)
{
    EmitIndirInstructionSize(0x32, regDst, ref memory);
}
````

Someone might use this to `builder.EmitXOR(Register.RAX, Register.RCX)` and this actually emitting `xor eax, ecx` (32 bit regs) is subtle.

Or just rename this to `EmitZeroReg` and drop the second parameter?

nattress replied:
I renamed to `EmitZeroReg` since that's what we really need in this PR. Thanks!

nattress committed:
Rename EmitXor -> EmitZeroReg

My `EmitXor` is not a fully general purpose Xor implementation. Rename it to make it clear what we're using the method for.

nattress commented:
@jkotas In fixing this bug I noticed that the [atypical callsite](https://github.com/dotnet/coreclr/pull/1413) support you added a few years ago looks unused (assuming LLILC is defunct). Think we can remove this to simplify the stub handling?

jkotas commented:
Yes, `CORINFO_ACCESS_ATYPICAL_CALLSITE` can be deleted.

I think we will want to add something similar to handle fast tailcalls in R2R, but it will need to have a different shape.

nattress committed:
Update ReadyToRunSymbolNodeFactory.cs

nattress commented:
Overall Crossgen2'd framework size reduced from 92,392,504bytes -> 87,051,832bytes (5.8%)

@trylek Mind reviewing this when you get a sec?

trylek reviewed (approved):
Looks great, thank you and sorry for the delay!

nattress commented:
/azp run runtime
